File: packages/engine/lib/execution/src/package/simulation/state/behavior_execution/package.py
import sys
import traceback

# TODO: Propagate field specs to runners and use in state and context objects
BEHAVIOR_INDEX_FIELD_KEY = '_PRIVATE_7_behavior_index'
BEHAVIOR_IDS_FIELD_KEY = '_PRIVATE_7_behavior_ids'


def _hash_behavior_id(lang_index, id_within_lang):
    # TODO: Either keep in sync with behavior id generation on Rust side of
    #       behavior execution package or change Rust side to generate
    #       single numbers (rather than pairs of numbers) in the first place.

    # We support fewer than 10 languages.
    return id_within_lang * 10 + lang_index


# `behavior_descs` should be a list of objects that have fields `id`, `name`, `source`, `columns`,
# `language` and `dyn_access`.
def _load_behaviors(behavior_descs):
    behaviors = {}
    warnings = []  # TODO: Accumulate warnings automatically with something like `hash_util.warn`
    for desc in behavior_descs:
        lang_idx = desc['id'][0]
        id_within_lang = desc['id'][1]

        if desc['language'] != "Python":
            behaviors[_hash_behavior_id(lang_idx, id_within_lang)] = {
                "name": desc['name'],
                "language": desc['language'],
            }
            continue

        try:
            # behavior_globals should contain a callable `behavior` if the user's code is correct
            behavior_globals = {}
            bytecode = compile(desc['source'], desc['name'], "exec")
            exec(bytecode, behavior_globals)
            behavior_fn = behavior_globals.get("behavior")

            if callable(behavior_fn):
                behaviors[_hash_behavior_id(lang_idx, id_within_lang)] = {
                    "name": desc['name'],
                    "language": desc['language'],
                    # TODO: Propagate desc['columns'] here so we can load
                    #       fewer columns based on behavior keys.
                    #       https://app.asana.com/0/1199548034582004/1201494318833036/f
                    "required_col_names": None,
                    "dyn_access": desc['dyn_access'],
                    "fn": behavior_fn
                }
            else:
                warnings.append(
                    f"Couldn't load behavior: No function named 'behavior': {desc['name']}"
                )

        except Exception:
            # Have to catch generic `Exception`, because user's code could throw anything.
            n_pkg_fns = 2
            e = str(traceback.format_exception(*sys.exc_info())[n_pkg_fns:])
            warnings.append(f"Couldn't load behavior `{desc['name']}`: {e}")

        # With the current implementation, failing to load a behavior
        # isn't an error if the behavior is never actually used. This
        # is consistent with Python usually not giving errors until an
        # erroneous piece of code is actually run.
        # TODO: Discuss whether making it an error would help user's debugging.

    return behaviors, warnings


def start_experiment(experiment, init_message, _experiment_context):
    experiment['behaviors'], warnings = _load_behaviors(init_message)
    return {
        "warnings": warnings
    }


def start_sim(experiment, sim, init_message, init_context):
    loaders = {
        BEHAVIOR_INDEX_FIELD_KEY: hash_util.load_full
    }
    return {
        "loaders": loaders,
    }


def _format_behavior_error(behavior_name, exc_info):
    n_pkg_fns = 2
    return f"Behavior `{behavior_name}` error: {traceback.format_exception(*exc_info)[n_pkg_fns:]}"


def _postprocess(agent_state):
    msgs = agent_state.messages
    if msgs is not None:
        for m in msgs:
            # Types are checked here because flush might not happen until
            # several Python behaviors later (due to behavior chaining)
            # and if type becomes correct in the meantime, the error won't
            # be raised at all, so whether an error is raised would depend
            # on the rest of the behavior chain, not just the current behavior.
            # TODO: OPTIM But maybe delay type-checking until flush after all
            #       if it has a big performance impact, since it wouldn't affect
            #       correct user code and isn't very likely to affect incorrect
            #       code.

            if isinstance(m["to"], str):
                m["to"] = [m["to"]]
            elif type(m["to"]) != list:
                raise TypeError(f"Message `to` field must be a list or a string: {m}")

            if not isinstance(m["type"], str):
                raise TypeError(f"Message `type` field must be a string: {m}")

    # TODO: OPTIM hasattr outside of loop over agents might be better than
    #       this try/except inside the loop.
    try:
        position = agent_state.position
        if position is not None:  # TODO: OPTIM Is position nullable?
            while len(position) < 3:
                position.append(0.0)
    except AttributeError:
        pass

    # TODO: OPTIM Same todos as for position
    try:
        direction = agent_state.direction
        if direction is not None:
            while len(direction) < 3:
                direction.append(0.0)
    except AttributeError:
        pass


# For each agent in the given group, execute behaviors and postprocess state,
# starting after the last behavior already executed (during this step / more generally
# behavior execution package call) and stopping when all behaviors are executed or
# the next behavior is in a different language (i.e. not Python).
def run_task(experiment, _sim, _task_message, group_state, group_context):
    next_lang = None
    agent_state = None
    agent_context = None

    for i_agent in range(group_state.n_agents()):
        # TODO: Reuse `agent_state` and `agent_context` objects.
        # agent_state = group_state.get_agent(i_agent, agent_state)
        # agent_context = group_context.get_agent(i_agent, agent_context)
        agent_state = group_state.get_agent(i_agent)
        agent_context = group_context.get_agent(i_agent)

        # ids of behaviors of this agent
        behavior_ids = getattr(agent_state, BEHAVIOR_IDS_FIELD_KEY)

        # `behavior_index` is the index of the first behavior that
        # hasn't been executed yet (during this step / package call).
        for i_behavior in range(int(agent_state.behavior_index()), len(behavior_ids)):
            # Behavior ids are shallow-loaded as an optimization, so
            # need to convert to a Python object here.
            # TODO: OPTIM Use `.values` attribute after upgrading Arrow.
            b_id = behavior_ids[i_behavior].as_py()
            behavior = experiment['behaviors'][_hash_behavior_id(b_id[0], b_id[1])]
            if behavior['language'] != "Python":
                # `next_lang` might be assigned to multiple times and overwritten,
                # but that's ok, because we can just use the last value that was
                # assigned to it.
                next_lang = behavior['language']
                break

            agent_state.set_dynamic_access(behavior['dyn_access'])
            try:
                behavior['fn'](agent_state, agent_context)
                _postprocess(agent_state)  # Errors from post-processing are considered user errors.
            except Exception:
                # Have to catch generic `Exception`, because user's code could throw anything.
                error = _format_behavior_error(behavior['name'], sys.exc_info())
                return {
                    "target": "Main",
                    "errors": [error]
                }

            # Increment the behavior index to point to the next one to be executed
            setattr(agent_state, BEHAVIOR_INDEX_FIELD_KEY, i_behavior + 1)

    return {
        "target": next_lang if next_lang is not None else "Main"
    }


def increment_behavior_index(agent_state, i_behavior):
    setattr(agent_state, BEHAVIOR_INDEX_FIELD_KEY, i_behavior + 1)

File: apps/engine/lib/execution/src/package/simulation/state/behavior_execution/package.js
// TODO: Propagate field specs to runners and use in state and context objects
const BEHAVIOR_INDEX_FIELD_KEY = "_PRIVATE_7_behavior_index";
const BEHAVIOR_IDS_FIELD_KEY = "_PRIVATE_7_behavior_ids";

// The `BEHAVIOR_INDEX_FIELD_KEY` column is reset to zero and the private `behavior_ids`
// column is written to
// in the simulation main loop immediately before running the behavior execution
// package.

// The `behaviors` column contains behavior names and can change in the
// middle of running the behavior execution package, but `__behaviors`
// contains behavior ids and shouldn't be modified.

const supported_languages = ["JavaScript", "TypeScript"];

const prepare_user_trace = (error, trace) => {
  let behavior_index = -1;
  for (var i = trace.length - 1; i >= 0; --i) {
    if (trace[i].isEval() && trace[i].getFunctionName() === "behavior") {
      behavior_index = i;
      break;
    }
  }

  if (behavior_index >= 0) {
    // Behavior function was found in stack trace.
    // Remove our (i.e. JS runner's) functions from end of trace.
    while (behavior_index < trace.length - 1) {
      // Behavior isn't last element.
      trace.pop();
    }
  }

  const frames = [];
  for (var i = 0; i < trace.length; ++i) {
    var t = trace[i];
    frames[i] = {
      file: t.getFileName(),
      // Behavior line numbers are off by 2 somehow.
      line: t.getLineNumber() - (i === behavior_index ? 2 : 0),
      fn: t.getFunctionName(),
    };
  }

  return {
    msg: error.toString(),
    frames: frames,
  };
};

/// `behavior_descs` should be a list of objects with fields `id`, `name`, `source`, `columns`,
/// `language` and `dyn_access`.
const load_behaviors = (experiment, behavior_descs) => {
  experiment.logged = "";
  const console = new Proxy(
    {},
    {
      get: (_, property) => {
        if (property === "log")
          return (...args) => {
            for (var i = 0; i < args.length; ++i) {
              experiment.logged = experiment.logged + args[i] + " ";
            }
            experiment.logged = experiment.logged + "\n";
          };

        return (...args) => {}; // generic noop function.
      },
      // TODO: non-noop implementation of `console` functions other than `log`
      // TODO: give error instead of returning noop function for methods that
      //       `console` object doesn't usually have
    },
  );

  const behaviors = {};
  for (var i = 0; i < behavior_descs.length; ++i) {
    const desc = behavior_descs[i];

    if (!supported_languages.includes(desc.language)) {
      behaviors[desc.id] = {
        language: desc.language,
      };

      continue;
    }

    const code = desc.source;

    let fn;
    try {
      fn = new Function(
        "hash_stdlib",
        "hstd",
        "console",
        `${code}\nreturn behavior`,
      )(hash_stdlib, hash_stdlib, console);
    } catch (e) {
      // Catch behavior code syntax errors and rethrow.
      Error.prepareStackTrace = prepare_user_trace;
      const trace = e.stack;
      trace.msg =
        "Couldn't load behavior (NAME " +
        desc.name +
        ", SOURCE " +
        code +
        "):" +
        trace.msg;
      throw new Error(JSON.stringify(trace));
    }
     
    let t = typeof fn;
    if (t !== "function") {
      throw new TypeError(
        desc.name + ": Behavior must be function, but is " + t + ".",
      );
    }

    behaviors[desc.id] = {
      fn: fn,
      name: desc.name,
      required_col_names: desc.columns,
      dyn_access: desc.dyn_access,
      language: desc.language,
      // Language of loaded behaviors is always Javascript,
      // since couldn't load them here otherwise.
    };
    // TODO
    // if (!desc.columns) {
    //     throw new TypeError(
    //         desc.name + " required_col_names " + typeof desc.columns
    //     );
    // }
  }

  experiment.behaviors = behaviors;
};

// Incorrect because behaviorIndex has historically been a function, not a property:
// const getters = {
//     "behaviorIndex": agent_state => agent_state[BEHAVIOR_INDEX_FIELD_KEY]
// }

export const start_experiment = (
  experiment,
  init_message,
  experiment_context,
) => {
  load_behaviors(experiment, init_message);
};

// Fill an array with a default value until its length is 3
const fill3 = (arr, val) => {
  while (arr.length < 3) {
    arr.push(val);
  }
};

const postprocess = (agent_state) => {
  const msgs = agent_state.messages;
  if (msgs) {
    for (var i = 0; i < msgs.length; ++i) {
      const m = msgs[i];
      if (typeof m.to === "string") m.to = [m.to];
    }
  }

  // Extend partial 3D coordinates with zeros.
  // TODO: Fill `scale`? (Fill with 1.0 instead of 0.0?) `rgb`?
  // `position` and `direction` are nullable fields.

  const position = agent_state.position;
  if (position) fill3(position);

  const direction = agent_state.direction;
  if (direction) fill3(direction);
};

export const run_task = (
  experiment,
  sim,
  task_message,
  group_state,
  group_context,
) => {
  // Reset `experiment.logged`, because it might have been written to
  // during a previous `run_task` call.
  experiment.logged = "";
  // Mutating `experiment.logged` here is ok because its value at the
  // start of a task doesn't depend on the order that tasks are executed
  // in (it's always "") and two tasks can't be executed simultaneously
  // in the same worker (they can be executed simultaneously in different
  // workers, but then the different workers would have their own instances
  // of the `experiment` object).

  let next_lang = null;
  let agent_state = null;
  let agent_ctx = null;

  const n_agents_in_group = group_state.n_agents();
  for (var i_agent = 0; i_agent < n_agents_in_group; ++i_agent) {
    // TODO: Reuse `agent_state` objects. When using the old `agent_state`, the indices for behaviors are borked
    agent_state = group_state.get_agent(i_agent, null);
    // TODO: Reuse `agent_ctx` objects. When using the old `agent_ctx`
    agent_ctx = group_context.get_agent(i_agent, null);

    const behavior_ids = agent_state[BEHAVIOR_IDS_FIELD_KEY];
    const n_behaviors = behavior_ids.length;
    for (
      var i_behavior = agent_state.behaviorIndex();
      i_behavior < n_behaviors;
      ++i_behavior
    ) {
      const b_id = behavior_ids.get(i_behavior);
      // We do this because behavior ids are shallow-loaded and
      // `b_id` is an Arrow Vec rather than a clean array
      const behavior = experiment.behaviors[[b_id.get(0), b_id.get(1)]];
      if (!supported_languages.includes(behavior.language)) {
        // TODO: A simple optimization would be to count the number of
        //       next-up behaviors in each language (other than JS) and
        //       (ignoring ties) choose the language with the most. This
        //       wouldn't hurt performance at all in the case where all
        //       behaviors are in JS.
        next_lang = behavior.language; // Multiple assignments are fine.
        break;
      }

      agent_state.set_dynamic_access(behavior.dyn_access);
      try {
        behavior.fn(agent_state, agent_ctx);
        postprocess(agent_state);
      } catch (e) {
        Error.prepareStackTrace = prepare_user_trace;
        const trace = e.stack;
        throw Error(JSON.stringify(trace));
      }

      // Increment the behavior index to point to the next one to be executed
      agent_state[BEHAVIOR_INDEX_FIELD_KEY] = i_behavior + 1;
    }
  }

  return {
    print: experiment.logged,
    target: next_lang || "Main",
    task: "{}", // TODO: Maybe this shouldn't be necessary
  };
};
